### PR TITLE
Feature/create compatible matches

### PR DESCRIPTION
**Pull request by LeandroSoria2**
### What changes does this PR include?
**Description**
https://trello.com/c/YPpFATmC/8-5-as-a-user-i-should-receive-notifications-when-a-compatible-target-has-been-created-so-we-can-start-getting-connected
**Summary**

- Add in service create_match
- Add RSpec test for a service

### CODE REVIEW

File: spec/services/target/create_target_services_spec.rb
Comment (juanmferreira93) [inline spec/services/target/create_target_services_spec.rb]:
```suggestion
      it 'creates a new match' do
```

File: app/services/targets/create_service.rb
Comment (JulianPasquale) [inline app/services/targets/create_service.rb]: Question about this: If the query returns multiple records, should we select the "best match" somehow (e.g. the one that is closer to the other target)?

File: spec/services/target/create_target_services_spec.rb
Comment (juanmferreira93) [inline spec/services/target/create_target_services_spec.rb]:
```suggestion
    context 'when there are not matches' do
```

File: spec/services/target/create_target_services_spec.rb
Comment (juanmferreira93) [inline spec/services/target/create_target_services_spec.rb]:
```suggestion
        it 'does not create a match' do
```

File: spec/services/target/create_target_services_spec.rb
Comment (juanmferreira93) [inline spec/services/target/create_target_services_spec.rb]:
```suggestion
    context 'when is out of range' do
```

File: spec/services/target/create_target_services_spec.rb
Comment (JulianPasquale) [inline spec/services/target/create_target_services_spec.rb]:
```suggestion
      it 'creates a match' do
```

File: spec/services/target/create_target_services_spec.rb
Comment (JulianPasquale) [inline spec/services/target/create_target_services_spec.rb]: Can we also test that this Match has the correct target ids?

File: spec/services/target/create_target_services_spec.rb
Comment (JulianPasquale) [inline spec/services/target/create_target_services_spec.rb]:
```suggestion
    context 'when two targets do not match' do
```

File: spec/services/target/create_target_services_spec.rb
Comment (JulianPasquale) [inline spec/services/target/create_target_services_spec.rb]:
```suggestion
      context 'when they have a different topic' do
```

File: spec/services/target/create_target_services_spec.rb
Comment (juanmferreira93) [reply spec/services/target/create_target_services_spec.rb]: this `context` should be inside the `context` defined in line 57, it's also a case for when there are no matches

File: spec/services/target/create_target_services_spec.rb
Comment (JulianPasquale) [inline spec/services/target/create_target_services_spec.rb]:
```suggestion
      it 'does not create a match' do
```

Comment (juanmferreira93) [inline spec/services/target/create_target_services_spec.rb]:
```suggestion
      end
      
      it 'does not create a match' do
```

File: spec/services/target/create_target_services_spec.rb
Comment (JulianPasquale) [inline spec/services/target/create_target_services_spec.rb]:
```suggestion
    context 'When they have the same user' do
```

Comment (juanmferreira93) [inline spec/services/target/create_target_services_spec.rb]:
```suggestion
    context 'when has the same user' do
```
again: this `context` should be inside the `context` defined in line 57, it's also a case for when there are no matches

File: spec/services/target/create_target_services_spec.rb
Comment (JulianPasquale) [inline spec/services/target/create_target_services_spec.rb]:
```suggestion
      it 'does not create a match' do
```

Comment (juanmferreira93) [inline spec/services/target/create_target_services_spec.rb]:
```suggestion
      end
      
      it 'does not create a match' do
```

File: spec/services/target/create_target_services_spec.rb
Comment (JulianPasquale) [inline spec/services/target/create_target_services_spec.rb]:
```suggestion
      it 'does not create a match' do
```

Comment (juanmferreira93) [inline spec/services/target/create_target_services_spec.rb]:
```suggestion
      it 'does not create a match' do
```

File: spec/services/target/create_target_services_spec.rb
Comment (juanmferreira93) [inline spec/services/target/create_target_services_spec.rb]: again: this `context` should be inside the `context` defined in line 57, it's also a case for when there are no matches

File: app/services/targets/create_service.rb
Comment (sebastiancaraballo) [inline app/services/targets/create_service.rb]: Avoid sending `params` from controllers as much as possible. Services should be clear interfaces of business behavior.

File: app/services/targets/create_service.rb
Comment (sebastiancaraballo) [inline app/services/targets/create_service.rb:24]: I don't think we need to use memoization here.

File: app/services/targets/create_service.rb
Comment (sebastiancaraballo) [inline app/services/targets/create_service.rb]:
```suggestion
        target = current_user.targets.create!(latitude: ..., longitude: ..., etc)
```

File: app/services/targets/create_service.rb
Comment (sebastiancaraballo) [inline app/services/targets/create_service.rb:33]: What happens if the user already has a match with the same target and/or person?

File: spec/services/target/create_target_services_spec.rb
Comment (sebastiancaraballo) [inline spec/services/target/create_target_services_spec.rb:63]: 👀

File: spec/services/target/create_target_services_spec.rb
Comment (sebastiancaraballo) [inline spec/services/target/create_target_services_spec.rb:78]: 👀

File: app/services/targets/create_service.rb
Comment (LeandroSoria2) [reply app/services/targets/create_service.rb:24]: This is used above, when the user can only create maximum targets

File: app/services/targets/create_service.rb
Comment (LeandroSoria2) [reply app/services/targets/create_service.rb]: I understand what is being planted, but in the card, there is nothing specified as to whether to match with the closest one, that's why the first one is chosen

File: spec/queries/target_query_spec.rb
Comment (juanmferreira93) [inline spec/queries/target_query_spec.rb]: is this spec ready?

File: app/models/target.rb
Comment (juanmferreira93) [inline app/models/target.rb]: you are not using this attribute in your code.. maybe this PR still on WIP?

Comment (LeandroSoria2) [reply app/models/target.rb]: Yes, this pr still on wip

Comment (JulianPasquale) [reply app/models/target.rb]: We should probably change this PR to draft then 🙂

File: app/queries/target_query.rb
Comment (AgostinaDev) [inline app/queries/target_query.rb:11]: So... basically, you're matching the target with any other target that's within the distance radius, with the same topic, unmatched and that does not belong to the same user that created the target you sent in the params. 

This means that `you'll possibly match with the same user multiple times with different targets` as long as they have targets with matched = false.

@juanmferreira93 I suppose you're playing the client role. Are you okay with ths definition?

I wouldn't be okay with this if this was a real project,  but if Juanma gives his thumbs up, I accept it.

File: spec/queries/target_query_spec.rb
Comment (AgostinaDev) [inline spec/queries/target_query_spec.rb]: I don't understand why you've defined new_target in all of these lines

File: spec/queries/target_query_spec.rb
Comment (AgostinaDev) [inline spec/queries/target_query_spec.rb]: like here

File: spec/queries/target_query_spec.rb
Comment (AgostinaDev) [inline spec/queries/target_query_spec.rb]: and here

File: spec/queries/target_query_spec.rb
Comment (AgostinaDev) [inline spec/queries/target_query_spec.rb]:
```suggestion
      context 'when there are not targets with the same topic' do
```

File: spec/queries/target_query_spec.rb
Comment (AgostinaDev) [inline spec/queries/target_query_spec.rb]:
```suggestion
        context 'when there are not targets without a match' do
```

File: spec/queries/target_query_spec.rb
Comment (AgostinaDev) [inline spec/queries/target_query_spec.rb]:
```suggestion
          it 'returns a complatible target' do
```

File: spec/queries/target_query_spec.rb
Comment (AgostinaDev) [inline spec/queries/target_query_spec.rb]:
```suggestion
        context 'when there are unmatched targets' do
```

File: spec/services/target/create_target_services_spec.rb
Comment (AgostinaDev) [inline spec/services/target/create_target_services_spec.rb]: I suggest a similar context structure to the one you have in `spec/queries/target_query_spec.rb`

File: spec/services/target/create_target_services_spec.rb
Comment (AgostinaDev) [inline spec/services/target/create_target_services_spec.rb]: Try to be consistent with the casing (always start the description with a lower casing)

File: spec/services/target/create_target_services_spec.rb
Comment (AgostinaDev) [inline spec/services/target/create_target_services_spec.rb]: Try to be consistent with the casing

File: app/queries/target_query.rb
Comment (juanmferreira93) [reply app/queries/target_query.rb:11]: @AgostinaDev, I understand your point, and it's a good one. In this opportunity I prefer to try to merge this PR as it is and then create a new ticket to improve that, taking in count that the ticket do not say anything about multiple matches (I could be wrong, but I think that is the case). 
As you mentioned, in a real project we should not merge that, but for this study case I think we are fine.
I think we could look into the `Match` table, for all related targets and his owners, then pluck the `id's` and add the condition in the TargetQuery.

File: app/controllers/api/v1/targets_controller.rb
Comment (AgostinaDev) [inline app/controllers/api/v1/targets_controller.rb]: 1. There's a typo here (target_attibutes -> `target_attributes`)
2. What are you using this for?

File: app/controllers/api/v1/targets_controller.rb
Comment (juanmferreira93) [inline app/controllers/api/v1/targets_controller.rb]: you defined it but never use it

File: app/services/targets/create_service.rb
Comment (AgostinaDev) [reply app/services/targets/create_service.rb:33]: @sebastiancaraballo check Juanma's answer to my question above (file app/queries/target_query.rb)

Comment (juanmferreira93) [reply app/services/targets/create_service.rb:33]: In this case for the same Target will not happened, because Lean flag the matched Targets to `matched: true` and the TargetQuery will not return it, but for the same person, as I told to Agos, we will allow it for now and will create a new ticket to fix it later.. to avoid block this PR so long.

File: db/migrate/20221010155749_add_matched_to_target.rb
Comment (juanmferreira93) [inline db/migrate/20221010155749_add_matched_to_target.rb]: you should add the default as false here

File: spec/queries/target_query_spec.rb
Comment (juanmferreira93) [inline spec/queries/target_query_spec.rb:23]: you are missing some cases here, like when the user is the same

File: spec/queries/target_query_spec.rb
Comment (juanmferreira93) [inline spec/queries/target_query_spec.rb:38]:
```suggestion
          end
          
          context 'when there are targers with same user' do
```

use empty lines to slit between `it's` and `context's` or generally to split ideas

File: spec/queries/target_query_spec.rb
Comment (juanmferreira93) [inline spec/queries/target_query_spec.rb:95]:
```suggestion
      end
      
      it 'does not return a compatible target' do
```

**Review by juanmferreira93 — APPROVED** [review]

**Review by sebastiancaraballo — APPROVED** [review]

**Review by AgostinaDev — APPROVED** [review]
Please address comments before merging

File: app/models/target.rb
Comment (AgostinaDev) [inline app/models/target.rb:33]: Nice 👍

File: spec/services/target/create_target_services_spec.rb
Comment (AgostinaDev) [inline spec/services/target/create_target_services_spec.rb:86]: I believe this context could be inside `context 'when there are targets without a match' do`,
at the same level `context 'when there are targets with different users' do` is

**Pull request merged**
